CRM/Contribute/BAO/Contribution.php

```diff
@@ -3974,34 +3974,25 @@ public static function validateFinancialType($financialTypeId, $relationName = '
    * @param int $participantId
    * @param bool $updateStatus
    *
-   * @return null|object
+   * @return int
+   *
+   * @throws \CRM_Core_Exception
+   * @throws \CiviCRM_API3_Exception
    */
   public static function recordAdditionalPayment($contributionId, $trxnsData, $paymentType = 'owed', $participantId = NULL, $updateStatus = TRUE) {
 
     if ($paymentType == 'owed') {
       $financialTrxn = CRM_Financial_BAO_Payment::recordPayment($contributionId, $trxnsData, $participantId);
+      if (!empty($financialTrxn)) {
+        self::recordPaymentActivity($contributionId, $participantId, $financialTrxn->total_amount, $financialTrxn->currency, $financialTrxn->trxn_date);
+        return $financialTrxn->id;
+      }
     }
     elseif ($paymentType == 'refund') {
       $trxnsData['total_amount'] = -$trxnsData['total_amount'];
-      $financialTrxn = CRM_Financial_BAO_Payment::recordRefundPayment($contributionId, $trxnsData, $updateStatus);
-      if ($participantId) {
-        // update participant status
-        // @todo this doesn't make sense...
-        $participantStatuses = CRM_Event_PseudoConstant::participantStatus();
-        $ids = CRM_Event_BAO_Participant::getParticipantIds($contributionId);
-        foreach ($ids as $val) {
-          $participantUpdate['id'] = $val;
-          $participantUpdate['status_id'] = array_search('Registered', $participantStatuses);
-          CRM_Event_BAO_Participant::add($participantUpdate);
-        }
-      }
+      $trxnsData['participant_id'] = $participantId;
+      return civicrm_api3('Payment', 'create', $trxnsData)['id'];
     }
```

Comment:
Just to be on the safe side if $payment NOT IN 'refund', 'owed' should we throw a fatal error message on ```else{...}``` loop?

Reply:
GIven we don't throw one now I don't think we should add one in 'just in case'

Reply:
ok no problem